File: bindings/java/Kryptic.java
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.Base64;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Kryptic Java client.
 *
 * Communicates with a running Kryptic server (python -m kryptic serve).
 * Requires Java 11+ (uses java.net.http.HttpClient — zero extra dependencies).
 *
 * Compile: javac Kryptic.java
 * Run:     java Kryptic
 */
public class Kryptic {

    private final String base;
    private final HttpClient http;

    public Kryptic(String host, int port) {
        this.base = "http://" + host + ":" + port;
        this.http = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(10))
            .build();
    }

    public Kryptic() {
        this("127.0.0.1", 7890);
    }

    // ── Low-level HTTP helpers ─────────────────────────────────────────────────

    private String post(String path, String jsonBody) throws Exception {
        HttpRequest req = HttpRequest.newBuilder()
            .uri(URI.create(base + path))
            .header("Content-Type", "application/json")
            .POST(HttpRequest.BodyPublishers.ofString(jsonBody, StandardCharsets.UTF_8))
            .build();
        HttpResponse<String> resp = http.send(req, HttpResponse.BodyHandlers.ofString());
        if (resp.statusCode() >= 400) {
            throw new RuntimeException("Kryptic error " + resp.statusCode() + ": " + resp.body());
        }
        return resp.body();
    }

    private String get(String path) throws Exception {
        HttpRequest req = HttpRequest.newBuilder()
            .uri(URI.create(base + path))
            .GET()
            .build();
        HttpResponse<String> resp = http.send(req, HttpResponse.BodyHandlers.ofString());
        return resp.body();
    }

    private String delete(String path) throws Exception {
        HttpRequest req = HttpRequest.newBuilder()
            .uri(URI.create(base + path))
            .DELETE()
            .build();
        HttpResponse<String> resp = http.send(req, HttpResponse.BodyHandlers.ofString());
        return resp.body();
    }

    private String extract(String json, String key) {
        Pattern p = Pattern.compile("\"" + key + "\"\\s*:\\s*\"([^\"]+)\"");
        Matcher m = p.matcher(json);
        return m.find() ? m.group(1) : "";
    }

    // ── Public API ─────────────────────────────────────────────────────────────

    public String health() throws Exception {
        return get("/health");
    }

    public Session session() throws Exception {
        String resp = post("/sessions", "{}");
        String sid = extract(resp, "session_id");
        if (sid.isEmpty()) throw new RuntimeException("No session_id in response: " + resp);
        return new Session(sid);
    }

    public String httpGet(String url) throws Exception {
        return post("/http/get", "{\"url\":\"" + url + "\"}");
    }

    public String httpPost(String url, String jsonBody) throws Exception {
        String body = String.format("{\"url\":\"%s\",\"json\":%s}", url, jsonBody);
        return post("/http/post", body);
    }

    // ── Session ────────────────────────────────────────────────────────────────

    public class Session {
        private final String id;

        Session(String id) {
            this.id = id;
        }

        public String goto_(String url) throws Exception {
            return post("/sessions/" + id + "/goto",
                "{\"url\":\"" + url + "\",\"wait_until\":\"domcontentloaded\"}");
        }

        public String title() throws Exception {
            String resp = get("/sessions/" + id + "/title");
            return extract(resp, "title");
        }

        public String html() throws Exception {
            String resp = get("/sessions/" + id + "/html");
            return extract(resp, "html");
        }

        public String currentUrl() throws Exception {
            String resp = get("/sessions/" + id + "/url");
            return extract(resp, "url");
        }

        public String text(String selector) throws Exception {
            String body = "{\"selector\":\"" + selector + "\"}";
            String resp = post("/sessions/" + id + "/text", body);
            return extract(resp, "text");
        }

        public void click(String selector) throws Exception {
            post("/sessions/" + id + "/click", "{\"selector\":\"" + selector + "\"}");
        }

        public void fill(String selector, String value) throws Exception {
            String body = String.format("{\"selector\":\"%s\",\"value\":\"%s\"}", selector, value);
            post("/sessions/" + id + "/fill", body);
        }

        public String evaluate(String js) throws Exception {
            String body = "{\"js\":\"" + js.replace("\"", "\\\"") + "\"}";
            String resp = post("/sessions/" + id + "/evaluate", body);
            return extract(resp, "result");
        }

        public byte[] screenshot() throws Exception {
            String resp = post("/sessions/" + id + "/screenshot", "{}");
            String data = extract(resp, "data");
            return Base64.getDecoder().decode(data);
        }

        public void blockResources() throws Exception {
            post("/sessions/" + id + "/block",
                "{\"resource_types\":[\"image\",\"stylesheet\",\"font\",\"media\"]}");
        }

        public void close() throws Exception {
            delete("/sessions/" + id);
        }
    }

    // ── Example main ───────────────────────────────────────────────────────────

    public static void main(String[] args) throws Exception {
        Kryptic k = new Kryptic();
        System.out.println("Health: " + k.health());

        Session s = k.session();
        s.blockResources();
        s.goto_("https://example.com");
        System.out.println("Title: " + s.title());
        System.out.println("H1: " + s.text("h1"));
        s.close();

        String httpResp = k.httpGet("https://httpbin.org/get");
        System.out.println("HTTP GET status line: " + httpResp.substring(0, 80));
    }
}

File: bindings/javascript/kryptic.js
/**
 * Kryptic JavaScript Client
 *
 * Communicates with a running Kryptic server (python -m kryptic serve).
 *
 * Install: no external dependencies — uses Node.js built-in `http` module.
 *
 * Usage:
 *   const { KrypticClient } = require('./kryptic');
 *
 *   const client = new KrypticClient(); // default: localhost:7890
 *
 *   // Browser session
 *   const session = await client.session();
 *   await session.goto('https://example.com');
 *   const title = await session.title();
 *   console.log(title);
 *   await session.close();
 *
 *   // HTTP-only (no browser)
 *   const resp = await client.httpGet('https://example.com');
 *   console.log(resp.status, resp.body.slice(0, 100));
 */

'use strict';

const http = require('http');

class KrypticError extends Error {
  constructor(message, status) {
    super(message);
    this.name = 'KrypticError';
    this.status = status;
  }
}

class KrypticClient {
  /**
   * @param {string} host  Kryptic server host (default: 127.0.0.1)
   * @param {number} port  Kryptic server port (default: 7890)
   */
  constructor(host = '127.0.0.1', port = 7890) {
    this.host = host;
    this.port = port;
  }

  _request(method, path, body = null) {
    return new Promise((resolve, reject) => {
      const payload = body ? JSON.stringify(body) : null;
      const options = {
        hostname: this.host,
        port: this.port,
        path,
        method,
        headers: {
          'Content-Type': 'application/json',
          ...(payload ? { 'Content-Length': Buffer.byteLength(payload) } : {}),
        },
      };

      const req = http.request(options, (res) => {
        let data = '';
        res.on('data', (chunk) => (data += chunk));
        res.on('end', () => {
          try {
            const parsed = JSON.parse(data);
            if (!parsed.ok) {
              reject(new KrypticError(parsed.error || 'Unknown error', res.statusCode));
            } else {
              resolve(parsed);
            }
          } catch (e) {
            reject(new KrypticError(`Invalid JSON response: ${data}`, res.statusCode));
          }
        });
      });

      req.on('error', reject);
      if (payload) req.write(payload);
      req.end();
    });
  }

  async health() {
    return this._request('GET', '/health');
  }

  /**
   * Create a new browser session.
   * @returns {Promise<KrypticSession>}
   */
  async session() {
    const { session_id } = await this._request('POST', '/sessions');
    return new KrypticSession(this, session_id);
  }

  async httpGet(url, headers = {}) {
    return this._request('POST', '/http/get', { url, headers });
  }

  async httpPost(url, { json = null, data = null, headers = {} } = {}) {
    return this._request('POST', '/http/post', { url, json, data, headers });
  }

  async httpBatch(urls) {
    const { results } = await this._request('POST', '/http/batch', { urls });
    return results;
  }
}

class KrypticSession {
  constructor(client, id) {
    this._client = client;
    this.id = id;
  }

  _post(action, body = {}) {
    return this._client._request('POST', `/sessions/${this.id}/${action}`, body);
  }

  _get(action) {
    return this._client._request('GET', `/sessions/${this.id}/${action}`);
  }

  async goto(url, waitUntil = 'domcontentloaded') {
    return this._post('goto', { url, wait_until: waitUntil });
  }

  async title() {
    const { title } = await this._get('title');
    return title;
  }

  async html() {
    const { html } = await this._get('html');
    return html;
  }

  async url() {
    const { url } = await this._get('url');
    return url;
  }

  async text(selector) {
    const { text } = await this._post('text', { selector });
    return text;
  }

  async click(selector) {
    return this._post('click', { selector });
  }

  async fill(selector, value) {
    return this._post('fill', { selector, value });
  }

  async evaluate(js) {
    const { result } = await this._post('evaluate', { js });
    return result;
  }

  async find(selector) {
    return this._post('find', { selector });
  }

  async screenshot(fullPage = false) {
    const { data } = await this._post('screenshot', { full_page: fullPage });
    return Buffer.from(data, 'base64');
  }

  async blockResources(resourceTypes = ['image', 'stylesheet', 'font', 'media']) {
    return this._post('block', { resource_types: resourceTypes });
  }

  async waitFor(selector, state = 'visible') {
    return this._post('wait_for', { selector, state });
  }

  async close() {
    return this._client._request('DELETE', `/sessions/${this.id}`);
  }
}

module.exports = { KrypticClient, KrypticSession, KrypticError };
